### ISSUE TITLE
bug: `re_extract` behavior differs between backends

### ISSUE THREAD

**Issue by jcrist** [ibis-project/ibis#4593]
The behavior of `re_extract` differs between backends, with regards to what the `index` parameter means.

This method has signature `re_extract(pattern, index)`, where `pattern` is a regex pattern with optional groups, and `index` is an index of the group to extract (returning `NULL` if no match or no group matches that index).

- `duckdb`, `postgres`, `clickhouse`, ...: `re_extract(pattern, 0)` returns the part matching the first group if there's a match, and `NULL` otherwise.
- `sqlite`, `pandas`, `dask`, `pyspark`: `re_extract(pattern, 0)` returns the _whole string_ if there's a match, and `NULL` otherwise. You need to pass in `1` not `0` to extract the first group.

To put it another way, given a column with a value `"row_one"`, `column.re_extract("row_([a-z])", 0)` returns `"one"` for backends in the first group, and `"row_one"` for backends in the second.

Given the docstring, I _think_ the first group has the intended behavior.

**Comment by saulpw** [comment]
Python `re.match()` also uses [group 0 to mean the entire match](https://docs.python.org/3/library/re.html#match-objects), so I propose that we use the sqlite/pandas approach.  Could we get duckdb/postgres to align by just wrapping the entire regex in an outer `()` for them?

**Comment by jcrist** [comment]
`regexp_extract` in duckdb ([docs](https://duckdb.org/docs/sql/functions/patternmatching)) already does that, we're offsetting the index by 1 to provide the current behavior. Some of the backends may be trickier to make return the full match for index 0 though, not sure.

I'm happy to go either way on this one, as long as we're uniform across backends. No strong thoughts.

**Comment by cpcloud** [comment]
+1 on matching the behavior of `re.match`.

**Comment by jcrist** [comment]
Sounds good! Since this would be a breaking change for some backends, it'd be good to get this in for 4.0.

**Comment by jcrist** [comment]
For rows where the pattern doesn't match at all, what should the return value be? NULL or the empty string?

Duckdb's native `regexp_extract` returns the empty string in both cases, and the current sql implementations in ibis look to be attempting to do the same. Postgres's native `regexp_match` function returns `NULL` in this case, so ibis currently uses a case statement to convert that to `""`.

**Comment by cpcloud** [comment]
> Is there a reason to support arbitrary values here, or could we restrict pattern and index to literal strings/integers?

A column of strings and a column of groups to extract might be a case where you'd want this.

I think we just handle it in the backend and raise an error during compilation. We do this elsewhere in the codebase.

I'm in favor of not restricting it unless it's breaking something to keep it that way.
